TheSadRogue.Primitives/PropertyChangedEventHelpers.cs

```diff
@@ -0,0 +1,161 @@
+        {
+            // Nothing to do; the value hasn't changed
+            if (propertyField is null)
+            {
+                if (newValue is null) return;
+            }
+            else if (propertyField.Equals(newValue)) return;
+
+            // Set new value and fire event
+            var oldValue = propertyField;
+            propertyField = newValue;
+            try
+            {
+                changedEvent?.Invoke(self, new ValueChangedEventArgs<TProperty>(oldValue, newValue));
+            }
+            catch (InvalidOperationException)
+            {
+                // If exception, preserve old value for future
+                propertyField = oldValue;
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Sets the given field to the new value, and fires the corresponding events.  The value will be properly
+        /// reverted to the old value if the Changed event handler throws InvalidOperationException.  The "changing"
+        /// event will be fired just _before_ the value actually changes; the "changed" event will be fired just after.
+        /// </summary>
+        /// <param name="self" />
+        /// <param name="propertyField">Field to set.</param>
+        /// <param name="newValue">New value to set to given field.</param>
+        /// <param name="changingEvent">Event to fire when change is about to occur.</param>
+        /// <param name="changedEvent">Event to fire after change occurs.</param>
+        /// <typeparam name="TObject">Type of the object the property resides on.</typeparam>
+        /// <typeparam name="TProperty">Type of the property.</typeparam>
+        public static void SafelySetProperty<TObject, TProperty>(this TObject self, ref TProperty propertyField, TProperty newValue,
+                                                                 EventHandler<ValueChangedEventArgs<TProperty>>? changingEvent,
+                                                                 EventHandler<ValueChangedEventArgs<TProperty>>? changedEvent)
+            where TObject : notnull
+        {
+            // Nothing to do; the value hasn't changed
+            if (propertyField is null)
+            {
+                if (newValue is null) return;
+            }
+            else if (propertyField.Equals(newValue)) return;
+
+            // Create event arguments
+            var args = new ValueChangedEventArgs<TProperty>(propertyField, newValue);
+
+            // Fire "pre-change" event
+            changingEvent?.Invoke(self, args);
+
+            // Set new value and fire event
+            var oldValue = propertyField;
+            propertyField = newValue;
+            try
+            {
+                changedEvent?.Invoke(self, args);
+            }
+            catch (InvalidOperationException)
+            {
+                // If exception, preserve old value for future
+                propertyField = oldValue;
+                throw;
+            }
```

Comment:
Does setting up try catch add overhead? If you're moving 200 entities all at once, would this be bad? I've never really researched that

Reply:
Tested this; there are some use cases for try-catch where it makes a difference; but this does not appear to be one of them:

## With try-catch
|                  Method |      Mean |     Error |    StdDev |
|------------------------ |----------:|----------:|----------:|
|            ChangedBasic |  8.978 ns | 0.1354 ns | 0.1200 ns |
| ChangingAndChangedBasic | 15.899 ns | 0.2278 ns | 0.2020 ns |

## Without try-catch
|                  Method |      Mean |     Error |    StdDev |
|------------------------ |----------:|----------:|----------:|
|            ChangedBasic |  9.279 ns | 0.1637 ns | 0.1532 ns |
| ChangingAndChangedBasic | 13.030 ns | 0.1651 ns | 0.1379 ns |